airflow/models/dag.py

```diff
@@ -800,10 +801,21 @@ def validate(self):
                 f"inconsistent schedule: timetable {self.timetable.summary!r} "
                 f"does not match schedule_interval {self.schedule_interval!r}",
             )
+        self.validate_executor_field()
         self.validate_schedule_and_params()
         self.timetable.validate()
         self.validate_setup_teardown()
 
+    def validate_executor_field(self):
+        for task in self.tasks:
+            if task.executor:
+                try:
+                    ExecutorLoader.lookup_executor_name_by_str(task.executor)
+                except AirflowException:
```

Comment:
We should make the exception in `lookup_executor_name_by_str` more specific and only catch that exception here. `AirflowException` is too broad.

Reply:
It's nice to have an exception at this level because we have access to the task id of the specific task that is causing the problem. I thought that would be nice to have from a user perspective. What do you think?

Reply:
I could also create custom `ExecutorException` or something along those lines to not throw such a broad exception.

Reply:
I agree that re-raising a precise exception here is best, since we have more context. Plus the lookup function is meant to be generic I think, so having that exception be more general is best.

Though I do think a ExecutorException is fine if we are worried about AirflowException (but I don't personally like creating custom exceptions very often).

Reply:
Agree with @uranusjr -> AirflowException is **NOT** supposed to be raised directly (even if it already is in a number of places). It's merely a parent exception for all the custom exceptions we implement.

But in this case I think we do not really **need** to have a custom exception. Python has surprisingly small number of built-in exceptions, but they are versatile enough to be used in most cases unless you want to have really specific exception you want to handle; https://docs.python.org/3/library/exceptions.html

Is there any reason we cannot use standard exception here (for example `ValueError`) or maybe one of already defined AirlfowException's ?

Reply:
To clarify (I think only Jarek correctly interpreted what I wanted to say…), I agree with catching the exception here to add more context, but I don’t like the fact this is catching AirflowException. The underlying function should be modified to raise a more specific (potentially new) exception that subclasses AirflowException, and this block here should only catch that specific exception class.

Reply:
I agree with @uranusjr, and I notice we sometimes even wrap an explicit exception and re-raise it with AirflowExecption in many places. We probably could address it more.

But in this case, if we want to catch the narrow exception, we probably will need to change https://github.com/apache/airflow/blob/b03aa6caa076171337054c3adafe74e775d92453/airflow/executors/executor_loader.py#L198 ?

Reply:
Sorry for the delay, I got caught up with some other tasks. These are all good points, thanks for weighing in! I've updated the `lookup_executor_name_by_str` method to throw a `ValueError` exception, which I catch here, and modify the error message to be more informative.